### PR TITLE
If T::get exists, define not_null<T>::get and return not_null<result_…

### PR DESCRIPTION
**Pull request by kkoenig**
Provides an example of how we could begin to treat `not_null<T>` as an alias for `T`

See comments on issue https://github.com/Microsoft/GSL/issues/89

### CODE REVIEW

File: include/gsl.h [hunk 1]
```diff
@@ -122,15 +122,15 @@ class not_null
...
-#ifdef _MSC_VER
-        __assume(ptr_ != nullptr);
-#endif
-        return ptr_;
-    } // the assume() should help the optimizer
-
-    operator T() const {  return get(); }
-    T operator->() const { return get(); }
+    // if T has .get() forward its value wrapped in not_null
+    template<typename U = T, typename = decltype(std::declval<U>().get())>
```
Comment (trebconnell) [inline include/gsl.h]: nit: most places in this code doing enable_if type things use the name Dummy for the template parameter

Comment (gdr-at-ms) [inline include/gsl.h]: I actually find NOT naming a parameter as a better indication that it is dummy.

Comment (trebconnell) [inline include/gsl.h]: I do too, just basing this off the rest of the code

Comment (neilmacintosh) [inline include/gsl.h]: In some cases, not naming a template parameter causes a problem for MSVC 2013, so the usage probably spread further than it needed to. If you are making changes and don't want the name and it doesn't cause a target compiler a problem....then feel free to do without it.

**Comment by JordanMaples** [comment]
Maintainer's call: sorry about the delay. This has been superseded as the Guidelines have evolved.
Closing.
Thanks

**Pull request closed without merging**